modules/collator-selection/src/lib.rs

```diff
@@ -406,30 +397,83 @@ pub mod pallet {
 	/// Play the role of the session manager.
 	impl<T: Config> SessionManager<T::AccountId> for Pallet<T> {
 		fn new_session(index: SessionIndex) -> Option<Vec<T::AccountId>> {
-			log::info!(
-				"assembling new collators for new session {} at #{:?}",
+			let candidates = Self::candidates()
+				.iter()
+				.map(|candidate| candidate.who.clone())
+				.collect::<Vec<_>>();
+			let result = Self::assemble_collators(candidates);
+
+			log::debug!(
+				"assembling new collators for new session {:?} at #{:?}, candidates: {:?}",
 				index,
 				<frame_system::Pallet<T>>::block_number(),
+				result,
 			);
 
-			let candidates = Self::candidates();
-			let candidates_len_before = candidates.len();
-			let active_candidates = Self::kick_stale_candidates(candidates.into_inner());
-			let active_candidates_len = active_candidates.len();
-			let result = Self::assemble_collators(active_candidates);
-			let removed = candidates_len_before - active_candidates_len;
+			//TODO
+			//frame_system::Pallet::<T>::register_extra_weight_unchecked(
+			//	T::WeightInfo::new_session(candidates_len_before as u32, removed as u32),
+			//	DispatchClass::Mandatory,
+			//);
 
-			frame_system::Pallet::<T>::register_extra_weight_unchecked(
-				T::WeightInfo::new_session(candidates_len_before as u32, removed as u32),
-				DispatchClass::Mandatory,
-			);
 			Some(result)
 		}
-		fn start_session(_: SessionIndex) {
-			// we don't care.
+
+		fn start_session(index: SessionIndex) {
+			<Candidates<T>>::mutate(|candidates| {
+				let mut candidates_inner = candidates.to_vec();
+				candidates_inner.iter_mut().for_each(|candidate| {
+					if candidate.validator {
+						<SessionPoints<T>>::insert(&candidate.who, 0);
+					} else {
+						candidate.validator = true;
+					}
+				});
+				*candidates = candidates_inner
+					.try_into()
+					.expect("Only modified existing elements of candidates_inner");
+			});
+
+			log::debug!(
+				"start session {:?} at #{:?}, candidates: {:?}",
+				index,
+				<frame_system::Pallet<T>>::block_number(),
+				<SessionPoints<T>>::iter().map(|(who, _)| who).collect::<Vec<_>>()
+			);
+
+			//TODO
+			//frame_system::Pallet::<T>::register_extra_weight_unchecked(
+			//	T::WeightInfo::start_session(candidates_len_before as u32, removed as u32),
+			//	DispatchClass::Mandatory,
+			//);
 		}
-		fn end_session(_: SessionIndex) {
-			// we don't care.
+
+		fn end_session(index: SessionIndex) {
+			let candidates_len_before = Self::candidates().len();
+			<SessionPoints<T>>::iter().for_each(|(who, point)| {
+				if point == 0 {
+					log::debug!(
+						"end session {:?} at #{:?}, remove candidate: {:?}",
+						index,
+						<frame_system::Pallet<T>>::block_number(),
+						who,
+					);
+
+					let outcome = Self::try_remove_candidate(&who);
```

Comment:
better to collect all the authors needs remove, and remove all of them in one go